R/reactive-cached.R

```diff
@@ -0,0 +1,384 @@
+#'     sliderInput("y", "y", 1, 10, 5),
+#'     "x * y: ",
+#'     verbatimTextOutput("txt")
+#'   ),
+#'   server = function(input, output) {
+#'     r <- cachedReactive(
+#'       # Use input$x and input$y for the cache key. This is a reactive
+#'       # expression.
+#'       list(input$x, input$y),
+#'       {
+#'         # The value expression is an expensive computation
+#'         message("Doing expensive computation...")
+#'         Sys.sleep(2)
+#'         input$x * input$y
+#'       }
+#'     )
+#'     output$txt <- renderText(r())
+#'   }
+#' )
+#'
+#'
+#' # Demo of using eventExpr with an actionButton
+#' shinyApp(
+#'   ui = fluidPage(
+#'     sliderInput("x", "x", 1, 10, 5),
+#'     sliderInput("y", "y", 1, 10, 5),
+#'     actionButton("go", "Go"),
+#'     div("x * y: "),
+#'     verbatimTextOutput("txt")
+#'   ),
+#'   server = function(input, output) {
+#'     r <- cachedReactive(
+#'       # Take a reactive dependency on input$go, but don't use its value
+#'       # for the cache key.
+#'       eventExpr = input$go,
+#'       # Use input$x and input$y for the cache key, but don't take a reactive
+#'       # dependency on them.
+#'       cacheKeyExpr = {
+#'         list(input$x, input$y)
+#'       },
+#'       {
+#'         message("Doing expensive computation...")
+#'         Sys.sleep(2)
+#'         input$x * input$y
+#'       }
+#'     )
+#'     output$txt <- renderText(r())
+#'   }
+#' )
+#'
+#'
+#' # Demo of cache key collisions and how to avoid them
+#' shinyApp(
+#'   ui = fluidPage(
+#'     numericInput("n", "n", 1),
+#'     "n * 2: ", verbatimTextOutput("txt2"),
+#'     "n * 3: ", verbatimTextOutput("txt3"),
+#'     "n * 4: ", verbatimTextOutput("txt4")
+#'
+#'   ),
+#'   server = function(input, output) {
+#'     r2 <- cachedReactive(input$n, {
+#'       input$n * 2
+#'     })
+#'     output$txt2 <- renderText(r2())
+#'
+#'     # BAD: This uses the same cache key as r2 above, and so it will collide with
+#'     # r2.
+#'     r3 <- cachedReactive(input$n, {
+#'       input$n * 3
+#'     })
+#'     output$txt3 <- renderText(r3())
+#'
+#'     # GOOD: By adding an ID string,this uses a different cache key from r2,
+#'     # and so its keys will not collide with those from r2.
+#'     r4 <- cachedReactive(list(input$n, "times4"), {
+#'       input$n * 4
+#'     })
+#'     output$txt4 <- renderText(r4())
+#'   }
+#' )
+#'
+#' }
+#'
+#' @export
+cachedReactive <- function(
+  cacheKeyExpr,
+  valueExpr,
+  ...,
+  eventExpr = NULL,
+  ignoreNULL = TRUE,
+  label = NULL,
+  domain = getDefaultReactiveDomain(),
+  cache = "app"
+) {
+  check_dots_empty()
+
+  # eventExpr is optional
+  eventExpr <- enquo(eventExpr)
+  if (identical(get_expr(eventExpr), NULL)) {
+    eventFunc <- NULL
+  } else {
+    eventFunc <- as_function(eventExpr)
+  }
+
+  label <- exprToLabel(substitute(cacheKeyExpr), "cachedReactive", label)
+
+  cacheKeyFunc <- as_function(enquo(cacheKeyExpr))
+
+  valueFunc <- as_function(enquo(valueExpr))
```

Comment:
Can we add a test that passes in quosures with non-default environments? (So it MUST eval in the quosure environment to retrieve the proper values.)

Reply:
Done in 77a8a783.